src/compiler/checker.ts

```diff
@@ -17088,11 +17089,43 @@ namespace ts {
                 }
             }
 
-            function isIdenticalTo(source: Type, target: Type): Ternary {
+            function traceUnionsOrIntersectionsTooLarge(source: Type, target: Type, reportErrors: boolean): void {
+                if (!tracing.isTracing()) {
+                    return;
+                }
+
+                if ((source.flags & TypeFlags.UnionOrIntersection) && (target.flags & TypeFlags.UnionOrIntersection)) {
+                    const sourceUnionOrIntersection = source as UnionOrIntersectionType;
+                    const targetUnionOrIntersection = target as UnionOrIntersectionType;
+
+                    if (sourceUnionOrIntersection.objectFlags & targetUnionOrIntersection.objectFlags & ObjectFlags.PrimitiveUnion) {
+                        // There's a fast path for comparing primitive unions
+                        return;
+                    }
+
+                    const sourceSize = sourceUnionOrIntersection.types.length;
+                    const targetSize = targetUnionOrIntersection.types.length;
+                    if (sourceSize * targetSize > 1E6) {
+                        if (reportErrors) {

```

Comment:
This probably doesn't need to be guarded by `reportErrors` anymore, no?

Reply:
I thought I should leave it so we didn't get a bunch of duplicate reports, but I may have misunderstood what the flag does.  I guess I should just test it.

Reply:
You can still get duplicate reports with it in place if a comparison happens to show up multiple times in a program, and we do do some comparisons with `reportErrors` off only.

Reply:
There are several duplicates without the check, but a bunch of valuable events are lost with it.  De-duping is easy, so I'll drop the check.